### PR TITLE
Add Sentry for error reporting

### PR DESCRIPTION
**Pull request by pepibumur**
### Short description 📝
Handled bugs and unhandled errors are not being reported at the moment. Therefore, we can only know about them if developers open issues on this repository.

### Solution 📦
Create a project on [Sentry](https://sentry.io) and set up Tuist to use it.

### Implementation 👩‍💻👨‍💻
- [x] Add Sentry.framework
- [x] Set up the project generation to link the `TuistCore` target against it.
- [x] Change the `ErrorHandler` logic to report bugs.
- [x] Observe unhandled errors and report them using the error handler.
- [x] Update the Rake `package` task to override the linking settings of the `tuist` and `tuistenv` binaries to link dynamically against `Sentry.framework`.
- [x] Pass the SentryDSN when building the Package for release.
- [x] Update the install logic in TuistEnv to pull the `Sentry.framework`
- [x] Update the install script to place the `Sentry.framework` next to `tuistenv`.
- [x] Document how to disable tracking.

**Setup code**
```swift
do {
        Client.shared = try Client(dsn: "xxx")
        try Client.shared?.startCrashHandler()
    } catch let error {
        print("\(error)")
    }
```

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -100,6 +135,16 @@ def storage
   )
 end
 
+def with_sentry_dsn
```
Comment (pepibumur) [inline Rakefile]: The sentry_dsn is a secret so this piece of code makes sure the value is defined in the `SentryDsn.swift` file when packaging a new release of Tuist.

File: Sources/TuistCore/Errors/ErrorHandler.swift [hunk 2]
```diff
@@ -57,6 +85,14 @@ public final class ErrorHandler: ErrorHandling {
...
             printer.print(errorMessage: message)
         }
+
         exiter(1)
     }
+
+    /// Returns true if the user decided to disable the analytics by setting the TUIST_ANALYTICS_DISABLED environment variable.
+    ///
+    /// - Returns: True if the analytics are disabled.
+    func isDisabled() -> Bool {
```
Comment (pepibumur) [inline Sources/TuistCore/Errors/ErrorHandler.swift]: cc @kwridan do you think this should be enough?

**Comment by pepibumur** [comment]
The PR is blocked by [this issue](https://github.com/getsentry/sentry-cocoa/issues/316). We need the Sentry SDK API to provide an interface to send all the events and subscribe to the completion of the delivery. That'll allow Tuist to wait until the events are reported, otherwise it exits before the events reach the server.

**Comment by pepibumur** [comment]
After some tests, I realized that the Swift Package Manager doesn't export the dSYMs and therefore the stacktraces in Sentry can't be desymbolicated. That complicates debugging the issue.

I suggest taking a different approach here. Instead of reporting the errors to Sentry, use the `GitHubClient` to report the errors as GitHub issues. What do you think @tuist/core ?

**Comment by kwridan** [comment]
Agreed, I think a simpler alternate is better suited here - from @ollieatkinson suggestions (from slack) we can start by catching / surfacing the errors in the command line that will aid better bug reports.

**Comment by ollieatkinson** [comment]
The suggestion on slack was to create a global function for tagging errors which can be propagated through standard swift error handling.

```swift
@discardableResult @inlinable
public func report<T>(_ block: @autoclosure () throws -> T, message: String = "", file: String = #file, line: Int = #line, function: String = #function) throws -> T {

    do {
        return try block()
    } catch let error as ReportedError {
        throw error
    } catch let error as FatalError {
        throw ReportedError(error: error, message: message, file: file, line: line, function: function)
    } catch let error {
        throw ReportedError(error: UnhandledError(error: error), message: message, file: file, line: line, function: function)
    }
    
}
```

We should expand on this further and provide support for capturing the output of the shell command which failed.

**Comment by pepibumur** [comment]
I'll open a PR following the suggested approach.

**Pull request closed without merging**